agent/ae/ae.go

```diff
@@ -96,6 +96,10 @@ type StateSyncer struct {
 	// syncChangesEvent generates an event based on multiple conditions
 	// when the state machine is performing partial state syncs.
 	syncChangesEvent func() event
+
+	// nextFullSync is the point in time when the next full sync should
+	// occur.
+	nextFullSync time.Time
```

Comment:
This is totally fine but one alternative way you could also have done this would be to have this be a `<-chan time.Time` and to initialize this to `time.After(s.Interval + s.stagger(s.Interval)` in the `resetNextFullSync` method. That way you never have to do the math on the times.

It being a chan would also potentially allow testing things more easily as you could send your own events into that channel.

Again, this is just FYI if you hadn't already considered it.

Reply:
Thanks, I changed it. I considered it before, but it didn't work out then because I was doing something different, but now it works!

I have another Question: right now the timer is reset *before* the full sync starts. Which means if the sync takes a long time like 2minutes, the next sync would start Interval-2minutes after that. 

Do you think that is an issue? I am wondering a bit about the semantics here. The way it was before was different, the timer started after the full sync.

Reply:
I think it is fine the way I did it.